src/geo_index.c

```diff
@@ -125,16 +103,31 @@ static t_docId *geoRangeLoad(const GeoIndex *gi, const GeoFilter *gf, size_t *nu
   return docIds;
 }
 
-IndexIterator *NewGeoRangeIterator(GeoIndex *gi, const GeoFilter *gf, double weight) {
-  size_t sz;
-  t_docId *docIds = geoRangeLoad(gi, gf, &sz);
-  if (!docIds) {
+IndexIterator *NewGeoRangeIterator(RedisSearchCtx *ctx, const GeoFilter *gf) {
+  GeoHashRange ranges[GEO_RANGE_COUNT] = {0};
+  double radius_meter = gf->radius * extractUnitFactor(gf->unitType);
+  calcRanges(gf->lon, gf->lat, radius_meter, ranges);
+
+  IndexIterator **iters = rm_calloc(GEO_RANGE_COUNT, sizeof(*iters));
+  ((GeoFilter *)gf)->numericFilters = rm_calloc(GEO_RANGE_COUNT, sizeof(*gf->numericFilters));
+  size_t itersCount = 0;
+  for (size_t ii = 0; ii < GEO_RANGE_COUNT; ++ii) {
+    if (ranges[ii].min != ranges[ii].max) {
+      NumericFilter *filt = gf->numericFilters[ii] =
+              NewNumericFilter(ranges[ii].min, ranges[ii].max, 1, 1);
+      filt->fieldName = rm_strdup(gf->property);
+      filt->geoFilter = gf;
+      struct indexIterator *numIter = NewNumericFilterIterator(ctx, filt, NULL, INDEXFLD_T_GEO);
+      if (numIter != NULL) {
+        iters[itersCount++] = numIter;
+      }
+    }
+  }
+  IndexIterator *it = NewUnionIterator(iters, itersCount, NULL, 1, 1);
```

Comment:
If itersCount is one there is no need to wrap it in a Union iterator. Also if its zero you can just return NULL no?